File: src/learning_gem5/hello_object.cc
#include "learning_gem5/hello_object.hh"

#include "base/trace.hh"
#include "debug/Hello.hh"

HelloObject::HelloObject(HelloObjectParams * params) :
    SimObject(params),
    event([this]{processEvent();}, name()),
    goodbye(params->goodbye_object),
    latency(params->time_to_wait),
    timesLeft(params->number_of_fires),
    myName(params->name)
{
    DPRINTF(Hello, "Created the hello object\n");
    panic_if(!goodbye, "HelloObject must have a non-null GoodbyeObject");
}

HelloObject*
HelloObjectParams::create()
{
    return new HelloObject(this);
}

void
HelloObject::processEvent()
{
    timesLeft--;
    DPRINTF(Hello, "Hello world! Processing the event! %d left\n", timesLeft);

    if (timesLeft <= 0) {
        DPRINTF(Hello, "Done firing!\n");

        goodbye->sayGoodbye(myName);
    } else {
        schedule(event, curTick() + latency);
    }
}

void
HelloObject::startup()
{
    schedule(event, latency);
}

File: src/learning_gem5/HelloObject.py
from m5.params import *
from m5.SimObject import SimObject

class HelloObject(SimObject):
    type = 'HelloObject'
    cxx_header = "learning_gem5/hello_object.hh"

    time_to_wait = Param.Latency("Time before firing the event")
    number_of_fires = Param.Int(10, "Number of times to fire the event"
    " before goodbye")

    goodbye_object = Param.GoodbyeObject("A goodbye object")

class GoodbyeObject(SimObject):
    type = 'GoodbyeObject'
    cxx_header = "learning_gem5/goodbye_object.hh"

    buffer_size = Param.MemorySize('1kB',
        "Size of buffer to fill with goodbye")
    write_bandwidth = Param.MemoryBandwidth('100MB/s', "Bandwidth to fill "
    "the buffer")
